File: 프로그래머스/lv2/87946. 피로도/피로도.py
from itertools import permutations

def calc(k, ordered_lst):
    result = 0
    for a, b in ordered_lst:
        if a > k or b > k:
            return result
        k -= b
        result += 1
    return result

def solution(k, dungeons):
    answer = 0
    for order in list(permutations(dungeons)):
        answer = max(answer, calc(k, order))
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> gDungeons;
vector<int> visited;
int answer;

// hp = 현재 남은 피로도 , count = 지금까지 돈 던전 수
void dfs(int hp, int count) {
    answer = max(answer, count);
    for (int i = 0; i < gDungeons.size(); i++) {
        if (visited[i]) {
            continue;
        }
        if (gDungeons[i][0] > hp) {
            continue;
        } 
        visited[i] = 1;
        dfs(hp - gDungeons[i][1], count+ 1);
        visited[i] = 0;
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    
    gDungeons = dungeons;
    visited = vector<int>(dungeons.size(), 0);
    dfs(k, 0);
    
    
    return answer;
}